### DISCUSSION TITLE
questions about "bulk attributions" system possibilities

### DISCUSSION THREAD

**Discussion by abubelinha** [bionomia/bionomia#346] [Q&A] [answered]
Just want to be sure I understand the way bulk attributions system works, and possible alternatives.

Looks like I should first download a .csv related to just one person, and then reupload it after filling in the `action` column plus at least `gbifID` (I am just dealing with other people data so "_not me_" doesn't apply to my use case).

So, if I understood correctly, this would only attribute that person (even if he/she collected the specimens with someone else), and both download/reupload should be done from the corresponding url;
`https://bionomia.net/help-others/{personID}/upload` 

Figure out I have 500 records which were collected by 6 people in the same trip:

1. Should I reupload the same data 6 times, using the url corresponding to each of their personal IDs?
2. No way to upload all that info at once in a GBIF-like DwC.csv style, with `gbifID` plus `recordedById` & `identifiedById` containing a separated list of personal IDs?

Now figure out the person I am attributing collected 58000 specimens during his/her life: 
- I would need 6 downloads due to maximum 10000 records file length. Would that limit apply only to downloads, or also to uploads?  In other words: can I merge all 58000 records in just one file for reuploading? (I'd only use `action` + `gbifID` columns, so it wouldn't be that big)
- Is there any problem if I just try to download stuff directly from GBIF, do reconciliation my own way, and try to upload the files to Bionomia? (one person at a time, unless you tell me there is a chance for option 2 above).

I know the suggested way is providing that info from GBIF DwC datasource `recordedById` & `identifiedById`.
But what if you are not the provider.
Or maybe you are but you just prefer to use Bionomia so your work (attributing other people) gets more visibility.

Thanks
@abubelinha

**Comment by dshorthouse** [comment, marked as answer]
> So, if I understood correctly, this would only attribute that person (even if he/she collected the specimens with someone else), and both download/reupload should be done from the corresponding url; `https://bionomia.net/help-others/{personID}/upload`

Yep, that's the way it works.
 
> Figure out I have 500 records which were collected by 6 people in the same trip:
 
> 1. Should I reupload the same data 6 times, using the url corresponding to each of their personal IDs?

Yes, that's correct. It's possible that some of the other links had already been made by another scribe in Bionomia. Per recipient, pre-existing attributions are skipped on upload. In other words, bulk upload does not do updates. Also, the candidate specimen download `.csv` excludes existing attributions per recipient such that repeat downloads (with intervening uploads/attributions) will be ever smaller with fewer and fewer records remaining to be bulk attributed to a particular recipient.

> 2. No way to upload all that info at once in a GBIF-like DwC.csv style, with `gbifID` plus `recordedById` & `identifiedById` containing a separated list of personal IDs?

Not at the moment, but it's something to consider as you suggested this time last year: https://github.com/bionomia/bionomia/issues/259#issuecomment-1455185680. No progress made on development for that yet.
 
> * I would need 6 downloads due to maximum 10000 records file length. Would that limit apply only to downloads, or also to uploads?  In other words: can I merge all 58000 records in just one file for reuploading? (I'd only use `action` + `gbifID` columns, so it wouldn't be that big)

The limit doesn't apply to the upload, but it's not particularly performant at the moment because it's synchronous. Better would be if upload were converted into a background task - one job per row - such that response would be immediate with a progress ticker.

> * Is there any problem if I just try to download stuff directly from GBIF, do reconciliation my own way, and try to upload the files to Bionomia? (one person at a time, unless you tell me there is a chance for option 2 above).

In principle, that'd work because only three columns are used at Bionomia's end: `gbifID`, `action`, and `not me` with your creator ORCID pulled from your session account and the recipient identifier pulled from the path in the URL. But, as mentioned above, download of candidate records in Bionomia is not comprehensive because it excludes pre-existing attributions. So, you may need to decide where it's most appropriate to launch your download if you have other purposes in mind than making attributions in Bionomia.

**Reply by abubelinha** [reply]
Thanks a lot.

In case I manage to do it, how much would you auggest to delay between uploads to not stress the server?

I.e. "upload up to 1000 records every X seconds"

**Reply by dshorthouse** [reply]
Good question. Uploads generally don't stress the server because these are converted into bulk inserts in batches of 250 at a time just as is often done via the user interface. However, this is still synchronous (for now) and so a browser timeout is perhaps more of an issue beyond 10,000 records in a single upload. Here's the ticket for converting/splitting an upload file into a background task such that response on the web form will be immediate and (with luck) you'd be presented with a progress ticker rather than waiting, waiting, and waiting for a large upload to complete: https://github.com/bionomia/bionomia/issues/348

**Reply by abubelinha** [reply]
That sounds great !

🤔 ... as a human. When that happens, would it be possible to keep current uploading system available somehow?
(url parameter, different address, whatever)

I am actually planning to upload from a Python script and I am OK with sending multiple small splitted files and wait for server responses.

I suspect  a progress bar ¿javascript? would make it quite more difficult for me to handle.

**Comment by dshorthouse** [comment]
> I am actually planning to upload from a Python script and I am OK with sending multiple small splitted files and wait for server responses.
> 
> I suspect a progress bar ¿javascript? would make it quite more difficult for me to handle.

Aha! I wondered if that might be your aim. Because any route under /help-others/ requires an active session (including uploads) with additional CSRF token checks, I suspect your Python script would fail to execute with a HTTP 401. What you'd want instead is a write API, which would mean that Bionomia first needs to have some form of authentication for machines, not just OAuth2 pass-thru for humans using ORCID. This mechanism requires a re-direct back to a receiving page where sessions are then launched in keeping with the CSRF token; that's impractical & unwieldy for machine-based scripts.

**Comment by abubelinha** [comment]
> What you'd want instead is a write API, ...

Yes we talked about that in [bionomia-api #3](https://github.com/bionomia/bionomia-api/issues/3), but doesn't make sense if there are no more users than me.
Nevermind. It's a bit hacky but I can authenticate in desktop browser, grab cookies/tokens from developer tools, and then recycle them to upload my .csv files generated from Python ... as long as current file-upload system keeps working when you finish #348
BTW I don't really see any long waitings for uploads.  Has anyone complained about that? 

A couple of things that I am missing a lot:
- **How can a scribe download-back all the attributions he/she has made?**  I am so interested in this since I use a mix of attribution systems (bulk upload for those datasets and users I closely know, web editor tables for claiming/unclaiming duplicate specimens of these users in other datasets).  This is quite a lot of work and I was assuming some kind of .csv backup to be available, but it is not. So I miss all my web editions. 
- Why doesn't Bionomia **let bulk uploads be used to update previous info? (at least those attributions made by the same user)**  
  What if scribes commit any mistakes? (i.e. upload 1000 attributions to the wrong user; would be easily corrected just reuploading them to the correct one). Not happened to me but just in case.
  
I would add these as a new feature requests, if you find them reasonable.
Thanks!

**Comment by dshorthouse** [comment]
> * **How can a scribe download-back all the attributions he/she has made?**  I am so interested in this since I use a mix of attribution systems (bulk upload for those datasets and users I closely know, web editor tables for claiming/unclaiming duplicate specimens of these users in other datasets).  This is quite a lot of work and I was assuming some kind of .csv backup to be available, but it is not. So I miss all my web editions.

This is the first time I've heard such a request. A number of scribes have made 1M+ attributions so this'd be quite a hefty download. 

> * Why doesn't Bionomia **let bulk uploads be used to update previous info? (at least those attributions made by the same user)**
>   What if scribes commit any mistakes? (i.e. upload 1000 attributions to the wrong user; would be easily corrected just reuploading them to the correct one). Not happened to me but just in case.

I'm not yet prepared to use "easily" as an adjective to describe the possible workflow. I'd assume for starters that it'd have to intelligently coordinate DELETE, PUT, and POST alongside an "are you sure?" confirmation.
 
> I would add these as a new feature requests, if you find them reasonable. Thanks!

Please do! I'd be curious to hear if others wish for the same.

**Comment by abubelinha** [comment]
> > - **How can a scribe download-back all the attributions he/she has made?** (...)
> 
> This is the first time I've heard such a request. A number of scribes have made 1M+ attributions so this'd be quite a hefty download.
> 

Indeed. But don't you see a contradiction here? 
If nobody is requesting this, then very few people will be using it.  So not a problem.
And most probably, those with 1M+ attributions scribes wouldn't be among them (if they had any interest, they would have requested this feature long time ago). 

Who could be interested then?  
Just those wanting to document history of biological collectors in a particular limited context (both taxonomic and geographical).  
I find Bionomia interface so useful for this purpose, because I can both make bulk attributions and then use fine-grained web interface to include/exclude some other suggested attributions.  And I'm so glad to use a shared resource like Bionomia instead of editing things locally and keeping them for me.

But it would be fair to recognize that to finish my work I need to get back from Bionomia all that info I shared. 
I am just meaning a 4 columns .csv [`action, gbifID, userID, not them`] (for any other info I could need, I may cope with GBIF and gbifIDs).
That would serve not just for my aforementioned purpose: it would also let me feed-in `recordedById`/`identifieById` when I republish to GBIF some of the involved datasets which I do curate.

As for the possible download size, I made a random test and this would only be a [767 KB zipped file for 100000 rows](https://github.com/bionomia/bionomia/files/14920828/bionomia_bulk_attribs_download_100000-fake-random-items.csv.zip) (using random gbifIDs and 100 random collector's ORCIDs). 

Probably less than that, taking in account the smaller length of wikidata ids. 
I also set `action`='recorded,identified' and `not them`='TRUE' for all records, which is an unreal situation (one of those columns will always be empty, and many times `action` value will take a single word).

Of course, this wouldn't need to be a live-download feature: a scheduled (multi-)compressed file creation would be just fine.
It could even be uploaded to Zenodo as a private deposit so it doesn't take any space in Bionomia at all, but the interested scribes still have access to their own work. I think that would be fair.


> > **let bulk uploads be used to update previous info?** 
> 
> I'm not yet prepared to use "easily" as an adjective to describe the possible workflow.

Sorry, my bad: "easily" was not meaning to describe the possible Bionomia's workflow.
I just intended to describe this feature would make _easy for scribes_ to correct their possible mistakes when doing bulk-uploads.
But I don't expect to commit such errors, so nevermind.

Just out of curiosity: does it never happen? (some scribe uploading a long .csv attributing to the wrong user)
How would you deal with that?

**Comment by dshorthouse** [comment]
> Indeed. But don't you see a contradiction here?

Not really because we could be talking cross-purposes without either of us knowing what someone may wish to do with the download and what might be the shortcomings in expectations.

> Who could be interested then? Just those wanting to document history of biological collectors in a particular limited context (both taxonomic and geographical).

Here's where there might be limitations. Had you earlier uploaded attributions for individual collectors, not all might later be returned, filtered for your own contributions. The reason for this is because pre-existing attributions trump a new upload - they're skipped. Likewise, someone may have later corrected the attributions you made via the user interface. A per-recipient download of specimens is comprehensive. However, a per-attributor download (i.e. your own attributions made) may not be comprehensive if you had expectations that _all_ your prior attributions be present.

Could Bionomia have instead been designed to accrue an ever-expanding list of attributor-recipient-action-gbifID combinations but just show the most recent one created in the event of multiple recipient-action-gbifID triplets? Perhaps. However, it was designed this way because at the time, gbifIDs were nowhere near as stable as they have been in the past 2 years.
 
> But it would be fair to recognize that to finish my work I need to get back from Bionomia all that info I shared. I am just meaning a 4 columns .csv [`action, gbifID, userID, not them`] (for any other info I could need, I may cope with GBIF and gbifIDs). That would serve not just for my aforementioned purpose: it would also let me feed-in `recordedById`/`identifieById` when I republish to GBIF some of the involved datasets which I do curate.

As above, it's the "all that info I shared" that Bionomia cannot do, given that some of them might have been later corrected (or removed) by others or did not stick had they been included in a bulk upload because the link already existed.
 
> Just out of curiosity: does it never happen? (some scribe uploading a long .csv attributing to the wrong user) How would you deal with that?

At the moment? It'd be a case by case basis, directly in the backend.

**Comment by abubelinha** [comment]
> Here's where there might be limitations. Had you earlier uploaded attributions for individual collectors, not all might later be returned, filtered for your own contributions

OK, now I see what you mean and clearly understand your doubts. 

Thanks for pointing all that out. Still, I am very interested in keeping "the incomplete remains of my own attributions":
1. When I started my uploads, I was aware of the "_pre-existing attributions are skipped_" system. 
2. Almost 100% attributions I am doing are the first and only attributions made to the people I helped. So I won't miss that much.
3. Because of 2, I can cope with missing items (and if somebody corrected me, then I was probably wrong in what I did, so not a big deal).

So, to be accurate, I am talking about downloading "only" those gbifID/userID/action combinations where I am "still" the scribe at the time of download generation.
Does this make more sense?
__________________
Let me also modify my above suggested 4-columns file format. 
The "not them" column info is quite valuable for Bionomia, but I'd rather suggest a 3-columns file format:

|action|gbifID|userID|
|---|---|---|
|recorded,identified|123456789|0000-0000-0000-0002|
|recorded|123456789|0000-0000-0000-0003|
|recorded|123456790|Q123456|
|identified|123456790|0000-0000-0000-0003|
|...|...|...|

|action|gbifID|userID|
|---|---|---|
|_not them_|123456789|0000-0000-0000-0001|
|_not them_|123456790|Q123123|
|...|...|...|

So this permits to separate "my" records in a two-set system, being the 1st the most important one for me to download.
I am not so interested in keeping my "_not them_" rows, which I'd only see useful in case I need to re-do all my Bionomia-curation job (hopefully not, since that would mean Bionomia lost that info for some reason).

This reduces again the file size (one column removed and quite a lot of rows).

I hope this can be done.
If that is too much hassle, may I instead email and ask for a one-time file generation from time to time?

Thanks!

**Reply by dshorthouse** [reply]
Ticket: https://github.com/bionomia/bionomia/issues/352
